connector/src/main/java/com/google/cloud/spark/bigquery/AvroSchemaConverter.java

```diff
@@ -0,0 +1,287 @@
+      return (getter, ordinal) -> Long.valueOf(getter.getShort(ordinal));
+    }
+    if (sparkType instanceof IntegerType && avroType.getType() == Schema.Type.LONG) {
+      return (getter, ordinal) -> Long.valueOf(getter.getInt(ordinal));
+    }
+    if (sparkType instanceof LongType && avroType.getType() == Schema.Type.LONG) {
+      return (getter, ordinal) -> getter.getLong(ordinal);
+    }
+    if (sparkType instanceof FloatType && avroType.getType() == Schema.Type.DOUBLE) {
+      return (getter, ordinal) -> Double.valueOf(getter.getFloat(ordinal));
+    }
+    if (sparkType instanceof DoubleType && avroType.getType() == Schema.Type.DOUBLE) {
+      return (getter, ordinal) -> getter.getDouble(ordinal);
+    }
+    if (sparkType instanceof DecimalType && avroType.getType() == Schema.Type.BYTES) {
+      DecimalType decimalType = (DecimalType) sparkType;
+      return (getter, ordinal) -> {
+        Decimal decimal = getter.getDecimal(ordinal, decimalType.precision(), decimalType.scale());
+        return DECIMAL_CONVERSIONS.toBytes(
+            decimal.toJavaBigDecimal(),
+            avroType,
+            LogicalTypes.decimal(decimalType.precision(), decimalType.scale()));
+      };
+    }
+    if (sparkType instanceof StringType && avroType.getType() == Schema.Type.STRING) {
+      return (getter, ordinal) -> new Utf8(getter.getUTF8String(ordinal).getBytes());
+    }
+    if (sparkType instanceof BinaryType && avroType.getType() == Schema.Type.FIXED) {
+      int size = avroType.getFixedSize();
+      return (getter, ordinal) -> {
+        byte[] data = getter.getBinary(ordinal);
+        if (data.length != size) {
+          throw new IllegalArgumentException(
+              String.format(
+                  "Cannot write %s bytes of binary data into FIXED Type with size of %s bytes",
+                  data.length, size));
+        }
+        return new GenericData.Fixed(avroType, data);
+      };
+    }
+    if (sparkType instanceof BinaryType && avroType.getType() == Schema.Type.BYTES) {
+      return (getter, ordinal) -> ByteBuffer.wrap(getter.getBinary(ordinal));
+    }
+
+    if (sparkType instanceof DateType && avroType.getType() == Schema.Type.INT) {
+      return (getter, ordinal) -> getter.getInt(ordinal);
+    }
+
+    if (sparkType instanceof TimestampType && avroType.getType() == Schema.Type.LONG) {
+      return (getter, ordinal) -> getter.getLong(ordinal);
+    }
+
+    if (sparkType instanceof ArrayType && avroType.getType() == Schema.Type.ARRAY) {
+      DataType et = ((ArrayType) sparkType).elementType();
+      boolean containsNull = ((ArrayType) sparkType).containsNull();
+
+      Converter elementConverter =
+          createConverterFor(et, resolveNullableType(avroType.getElementType(), containsNull));
+      return (getter, ordinal) -> {
+        ArrayData arrayData = getter.getArray(ordinal);
+        int len = arrayData.numElements();
+        Object[] result = new Object[len];
```

Comment:
We are allocating the Object array in the inner loop. This may cause a lot of allocations and will likely result in GC. Is there a way to avoid this by doing this lazily?

Reply:
We are allocating the result array - this array is then wrapped by Arrays.asList(), serving as the backing data structure of the list. This code is a direct 1:1 port of the spark-avro scala code